pkg/queue/breaker.go

```diff
@@ -99,60 +137,66 @@ type Semaphore struct {
 	mux      sync.Mutex
 }
 
-// Acquire receives the token from the semaphore, potentially blocking
+// Acquire receives the token from the semaphore, potentially blocking.
 func (s *Semaphore) Acquire() {
 	<-s.queue
 }
 
-// Release releases the token to the queue
-// The operation is potentially blocking when the queue is full
-func (s *Semaphore) Release() {
-	s.AddCapacity(1)
+// Release potentially puts the token back to the queue.
+// If the semaphore capacity was reduced in between and is not yet reflected,
+// we remove the tokens from the rotation instead of returning them back.
+func (s *Semaphore) Release() error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.reducers > 0 {
+		s.capacity -= s.reducers
+		s.reducers--
+	} else {
+		// We want to make sure releasing a token is always non-blocking.
+		select {
+		case s.queue <- s.token:
+		default:
+			// this should never happen
+			return ErrRelease
+		}
+	}
+	return nil
 }
 
-// ReduceCapacity removes tokens from the rotation
-// It tries to acquire as many tokens as possible, if there are not enough tokens in the queue,
-// it postpones the operation for the future by increasing the `reducers` counter
-func (s *Semaphore) ReduceCapacity(size int32) error {
+// AddCapacity increases the number of tokens in the rotation.
+// Increases only if the size is positive, does nothing otherwise.
+// An error is returned if not all capacity could be added to the semaphore.
+// This error could happen when the number of tokens exceeds the semaphore's queue size.
```

Comment:
This logic feels slightly strange since, consider two situations.
MaxCapacity = 5
All 5 tokens in flight
Add(5) --> all  get added.

Alternatively:

MaxCapacity = 5
No tokens in flight
Add(5) --> none gets added.

Reply:
By `MaxCapacity` you mean the channel size of the semaphore?
The size should be > then the number of endpoints that must be added. Otherwise one gets an error.

Reply:
Yes, MaxCapacity == channel buffer size.
What I mean that success of this function depends on the state of semaphore (i.e. how many items are in flight right now). If that's the desired behavior, it might be fine, but we probably need better comment around it?

Reply:
I think I didn't state it explicitly in the comments to the semaphore definition. Will add some more details to that. 
The behaviour is indeed intended, it was a trade-off to reuse the existing implementation of the breaker with minimal changes. The assumption is that we set the size of the channel high enough so it much more then the number of the endpoints.